company-etags.el

```diff
@@ -74,15 +89,176 @@ Set it to t or to a list of major modes."
           (setq company-etags-buffer-table (company-etags-find-table))
         company-etags-buffer-table)))
 
+(defmacro company-etags-push-tagname (tagname tagname-dict)
+  "Push TAGNAME into TAGNAME-DICT."
+  `(let* ((c (elt ,tagname 0)))
+    (cond
+     ((or (and (>= c ?a) (<= c ?z))
+          (and (>= c ?A) (<= c ?Z))
+          (eq c ?$)
+          (eq c ?#)
+          (eq c ?@)
+          (eq c ?%)
+          (eq c ?_)
+          (eq c ?!)
+          (eq c ?*)
+          (eq c ?&)
+          (and (>= c ?0) (<= c ?9)))
+      (push ,tagname (gethash c ,tagname-dict)))
+     (t
+      (push ,tagname (gethash ?' ,tagname-dict))))))
+
+(defun company-etags-extract-tagnames (text)
+  "Extract tag names from TEXT."
+  (let* ((pattern "\177\\([^\177\001\n]+\\)\001")
+         (start 0)
+         (tagname-dict (make-hash-table))
+         (i 0))
+
+    ;; initialize hashtable whose key is from a...z and A...Z
+    (while (< i 26)
+      ;; make sure the hash value is not nil
+      (puthash (+ ?a i) '() tagname-dict)
+      (puthash (+ ?A i) '() tagname-dict)
+      (setq i (1+ i)))
+
+    ;; initialize hashtable whose key is from 0...9
+    (setq i 0)
+    (while (< i 10)
+      ;; make sure the hash value is not nil
+      (puthash (+ ?0 i) '() tagname-dict)
+      (setq i (1+ i)))
+    ;; other key used as the first character of variable name
+    (puthash ?$ '() tagname-dict)
+    (puthash ?_ '() tagname-dict)
+    (puthash ?# '() tagname-dict)
+    (puthash ?& '() tagname-dict)
+    (puthash ?@ '() tagname-dict)
+    (puthash ?! '() tagname-dict)
+    (puthash ?* '() tagname-dict)
+    (puthash ?% '() tagname-dict)
+    ;; rubbish bin
+    (puthash ?' '() tagname-dict)
+
+    (while (string-match pattern text start)
+      (let* ((m (match-string 1 text)))
+        (company-etags-push-tagname m tagname-dict)
+        (setq start (+ 4 (match-end 0)))))
+    tagname-dict))
+
+(defun company-etags-append-new-tagname-dict (new-tagnames file-info)
+  "Append NEW-TAGNAMES to FILE-INFO."
+  (dolist (tagname new-tagnames)
+    (company-etags-push-tagname tagname (plist-get file-info :tagname-dict))))
+
+(defun company-etags-all-completions (prefix tagname-dict)
+  "Search for partial matches to PREFIX in TAGNAME-DICT."
+  (let* ((c (elt prefix 0))
```

Comment:
Does this handle of case of prefix being the empty string?

Reply:
fixed